### ISSUE TITLE
Mid-traversal GraphStep are not optimized away

### ISSUE THREAD

**Issue by JPMoresmau** [pietermartin/sqlg#110]
```
this.sqlgGraph.traversal().V().hasLabel("A").as("a").V().hasLabel("B").as("b").select("a", "b");
```

When reaching the second V(), we do not optimize it (and the following) into a new SgqlGraphStepCompiled, so we do select * on all types, which is very inefficient...

**Comment by pietermartin** [comment]
4 tests on `SqlgGraphPostgresProcessStandardTest` are failing.

The one I started investigating is `VertexTest.g_V_hasLabelXpersonX_V_hasLabelXsoftwareX_name`

I have duplicated it in `MidTraversalGraphTest.g_V_hasLabelXpersonX_V_hasLabelXsoftwareX_name` to make testing easier.

**Comment by pietermartin** [comment]
For the query
```
this.sqlgGraph.traversal().V().hasLabel("A").as("a").V().hasLabel("B").as("b").select("a", "b");
```
if there are many `A`s then for each A the `V().hasLabel("B")` will have to execute. This is not happening at the moment and its not easy either as SqlgGraphStepCompiled consumes its `ReplacedSteps` so the `SqlgGraphstepStrategy` will have to run again.

However this is also a tad silly as it will be the same query over and over, so perhaps in this case we must think of resetting the `ResultSet` cursor to the beginning and just reiterate.
WDYT?

**Comment by JPMoresmau** [comment]
The two things are related. The issue is that in SqlgGraphStepCompiled we have an iterator over the result, that we never reset, while in GraphStep the iterator can correctly reset. So if I capture the iterator result into a list and then replay it when we reach the next previous head, it fixes the bug and I suppose it also has the nice side effect that we do not rerun the SQL.

**Comment by pietermartin** [comment]
Yes that will do.

Currently the jdbc also caches the full `ResultSet`, its the default behavior. There is a way to to tell the driver to lazily fetch result. Its a //TODO, mostly need to check the performance and memory impact. 

However seeing as its fully cached already I thought we could reset the `ResultSet` (also needs some setting for that to be possible and there is the problem making sure we do not leak resources). That way we will not cache the same data twice.

All that said I'd go with your way first and make it work. Then later we can optimize memory if its worthwhile and try the other way.

**Comment by pietermartin** [comment]
fixed by #114
